### ISSUE TITLE
Thousands of hanging /rt/connect?transport=websockets requests on server

### ISSUE THREAD

**Issue by guylando** [aspnet/SignalR#1980]
moving here issue https://github.com/SignalR/SignalR/issues/4095 as suggested by @halter73 

While trying to improve our servers performance, I discovered that there are thousands of hanging /rt/connect?transport=websockets and also /rt/reconnect?transport=websockets on our productions server.
Some of the requests have "Time Elapsed" of more than a day.
They all have ExecuteRequestHandler state AspNetCoreModule module name.
We are using a kestrel net core website behind an iis on windows server 2016.
Is this the expected behavior? Because I am sure it causes performance problems if the server is not being restarted for a few days.
Also in case this is not the correct behavior, is there some configuration to limit the length of the requests? if the requests time will be limited it won't hurt anything?

**Comment by anurse** [comment]
From your comment over on the original issue:

> websocket connections must have the rt/connect or rt/reconnect request active as long as they work? 

This is correct. WebSockets are persistent connections. That's the reason they exist.

> this limits the amount of concurrent websocket connections and for example

Yes, you are limited by the number of concurrent connections you can have active on your server. 

> non-server window it allows only 10 concurrent clients.

I'm not familiar with any limitation as low as 10. IIS does have a concurrent connection limit and you'll have to manage that as per the IIS documentation. Can you clarify the scenario in which you are seeing an issue?

Finally, can you provide the names and versions of SignalR packages you are using?

**Comment by anurse** [comment]
Also from your earlier comment:

> we use Microsoft.AspNetCore.SignalR.Server (waiting for .net core 2.1 to upgrade) on server and signalr 2.2.2 on client.

These versions are not supported. They were prototype versions and never intended to be used in production environments. The only supported version of SignalR for ASP.NET Core is `Microsoft.AspNetCore.SignalR 1.0.0-preview2-final`

**Comment by guylando** [comment]
reference on the small limitation on windows 7\10 here:
https://github.com/SignalR/SignalR/issues/3741#issuecomment-232873380
http://www.jpelectron.com/sample/WWW%20and%20HTML/IIS-%20OS%20Version%20Limits.htm

we wanted our qa to test version on the iis of our developer and encountered a limit which prevented us from opening more than 2-3 tabs (each having several websockets and in total 10)

also the thousands on hanging requests on the server (hundreds from same ip) look scary regarding the consequence on our servers performance and users amount it can support.

we use Microsoft.AspNetCore.SignalR.Server (waiting for .net core 2.1 to upgrade) on server and signalr 2.2.2 on client.

would be glad to hear advise on how to make those thousands of connections close automatically after some time and prevent the situation we see, and without breaking valid clients connections.
specifically prevents a single ip having hundreds of hanging connections

**Comment by guylando** [comment]
we use net core 2.0.1 in all packages. is the signalr version you suggested compatible with it?

**Comment by anurse** [comment]
> we use net core 2.0.1 in all packages. is the signalr version you suggested compatible with it?

No. SignalR is part of ASP.NET Core 2.1, so it requires the latest previews of .NET Core 2.1 and ASP.NET Core 2.1. Previous **prototype versions** were compatible with older versions of .NET Core because they were written before .NET Core 2.1.

> also the thousands of hanging requests on the server (hundreds from same ip) look scary regarding the consequence on our servers performance and users amount it can support.

This is always something you need to consider when adding persistent connections to an application. Each connection will consume some resources on the server, you need to decide how you want your server to behave when resources run low. Every SignalR connection is a persistent connection that remains open until either the client or the server closes it. It's incorrect to consider these "hanging requests", since the entire purpose of these connections is to be persistent until either party decides to terminate it.

> would be glad to hear advise on how to make those thousands of connections close automatically after some time and prevent the situation we see, and without breaking valid clients connections.
> specifically prevents a single ip having hundreds of hanging connections

In the latest previews, you have the ability to abort a connection from within a Hub by calling `Context.Abort()`, you can use this to terminate misbehaving connections. You could have a background task that tracks all the connections (you can register connections with this tracker in the `OnConnectedAsync` virtual method in your Hub), and terminates them using `Abort` after some timeout, or if they haven't sent/received any messages for some time.

**Comment by anurse** [comment]
> in SignalR/SignalR#3741 (comment) he said that the concurrent requests were caused by long polling fallback and suggested to only allow websockets so I assumed that the websockets didnt influence the requests limitation, that's why I was wondering on this (and it contradicts what you said unless I misunderstood what he said)

A WebSocket request is like any other HTTP request. If there are connection limits that restrict concurrent HTTP connections, a WebSocket is no different from Long Polling. I'm not sure what the author of that comment was referring to.

**Comment by guylando** [comment]
thanks a lot for your answers, you cleared some things to me. Just to finish this, I would like to hear your opinion on the following: currently each signalr connection from the client to the server creates a task in the server which runs an infinite loop on the server with some configurable sleeping period and when server has update to client then he sends the update and otherwise sleeps until next loop round. Is this the standard way of implementing server handling of signalr? 
some of the info which should trigger a request to the client comes from the mvc controller. Is it possible to somehow trigger signalr client function from inside mvc controller? because then we could get read of the infinite loop and just trigger it on demand. However it raises the problem of maybe the signalr connection dying if the infinite loop finishes? Would be glad to hear your opinion on this since maybe those infinite loops are also a factor in the performance issue

**Comment by anurse** [comment]
> currently each signalr connection from the client to the server creates a task in the server which runs an infinite loop on the server with some configurable sleeping period and when server has update to client then he sends the update and otherwise sleeps until next loop round.

Yeah, don't do that 😄 . Use `IHubContext` as described below.

> Is it possible to somehow trigger signalr client function from inside mvc controller

Yes! If you inject `IHubContext<YourHubClass>` into your MVC controller via DI, you'll find that it has methods on it that can be used to invoke methods on the client.

> However it raises the problem of maybe the signalr connection dying if the infinite loop finishes?

SignalR keeps the connection alive until either the client disconnects or the server calls Abort. There's no need for you to loop anywhere. It would help a bit if I could see the basic structure of your Hub code to see if I understand what you're trying to describe.

> if websocket is basically the same as long polling - a long http request which passes data through it, then why is websocket any better than long polling?

WebSockets are definitely not the same as Long Polling 😄 . In Long Polling, the client issues a request for new messages, but the server holds the request until data is available. Once a message is ready, the poll request completes and the server sends the data. The client then has to issue a new **request** before any new messages can be retrieved. Long Polling is simply an optimization on the general idea of Polling, where the client simply keeps repeatedly asking for new messages, until one arrives. There is also a lot of network traffic and overhead in sending each message to the client because the client has to send a complete HTTP request and the server has to send a complete HTTP response to carry the message. For the server to "send" a message to the client it has to **wait** for the client to **request** the message.

In WebSockets, an initial HTTP request is sent by the client to "Upgrade" the connection to a WebSocket. When the server accepts the upgrade, the connection is left open, and data is transferred freely between server and client using an efficient binary protocol. When the server wishes to send a message, it simply sends the data, resulting in a much lower latency. After the initial handshake, a WebSocket is no different than a standard Socket connection exchanging raw bytes, hence the name.

Long Polling is the "most compatible" mechanism for sending messages from the server to the client. Every Web Browser supports it, so it is a useful fallback. WebSockets is the "most efficient" mechanism for sending messages from server to client (and client to server), but not all browsers support it (though all modern ones do) and some intermediate proxies or servers cause issues. SignalR's value is that we automatically detect the capabilities of the browser, proxy and server involved and choose an appropriate transport automatically, so you don't have to worry about it.

**Comment by guylando** [comment]
Thanks for the detailed answer! 

Could you throw a few words also about ServerSentEvents? because it always appears as something between longpolling and websockets in the priority.

Regarding our code, the current situation is that there is a static thread safe dictionary saving tasks per connection id and we have onConnected code to create a new task for the connection and put it in the dictionary, onRemove aborts the task of the connection id and removes it from the dictionary, and onReconnect restarts the task.

We will utilize your advise on IHubContext for situations where an mvc action should trigger a UI update, Thanks!

However there are a lot of other situations in our system where some backend process performes changes on the data and those changes should trigger a change in the clients UI. In this situation our current infinite loop which monitors the database discovers the changes and triggers the clients UI change. Also I have seen in your example you use an infinite loop also for similar situation for stock ticker here: https://github.com/aspnet/SignalR-samples/blob/master/StockTickR/StockTickRApp/StockTicker.cs#L57
So in those situations the infinite loop with a task.delay is the best practise? Or is there some other trick which could help do this on demand? Or is it our own job to build a mechanism (for example some interprocess communication anouncing the change, but even then, how will the code which was triggered by the IPC will have access to the hub? is the IHubContext available everywhere all the time? even in singleton background tasks inside the site? If not then what do you suggest for this scenario.

**Comment by anurse** [comment]
> Could you throw a few words also about ServerSentEvents? because it always appears as something between longpolling and websockets in the priority.

Yep! Server-Sent Events is like a one-way WebSocket. The client establishes a connection to the server, and the server can stream back data one message at a time without requiring a new request. This is basically what WebSockets does **except** WebSockets also allows the **client** to send messages over that connection but Server-Sent Events does **not**. So it's better than Long Polling because the server can send messages without polling, but the client still has to use HTTP POST requests to send messages (which has a lot of overhead). Server-Sent Events is supported on Chrome and Firefox, but not Internet Explorer or Edge. So the transports break down like this:

| Transport | Chrome* | Firefox | Edge | IE | Safari** | Notes |
| ---- | ----- | ----- | ----- | ----- | ----- | ----- |
| Long Polling | Yes | Yes | Yes | Yes | Yes | Slowest, but most compatible |
| Server-Sent Events | Yes | Yes | **No** | **No** | Yes | Better, but less compatible |
| WebSockets | Yes | Yes | Yes | Yes | Yes | Best, but some servers/proxies interfere |

\* Includes Android
\*\* Includes iOS

Honestly, most modern applications on modern browsers will use WebSockets as it is broadly supported and even though I warned that proxies sometimes interfere, almost all modern ones do not interfere (particularly if you use HTTPS, since they have no way of doing so).

> is the IHubContext available everywhere all the time?

Yes! `IHubContext` is in DI and available in any DI-injected service.

> Regarding our code, the current situation is that there is a static thread safe dictionary saving tasks per connection id and we have onConnected code to create a new task for the connection and put it in the dictionary, onRemove aborts the task of the connection id and removes it from the dictionary, and onReconnect restarts the task.

I'd remove all that and use IHubContext. It's exactly what it's designed for.

> So in those situations the infinite loop with a task.delay is the best practise?

The sample you linked is just designed to show how the streaming APIs work, it's not necessarily a Real World example. The ideal situation is for your app to receive some kind of notification when there is new data and then use `IHubContext` to pass it along. Any time you're looping, checking if there's new data, and then sleeping, you're making the processor do a bunch of work for no benefit. Obviously sometimes it's the only way to make things work, but if you can avoid it, I strongly suggest you do.

**Comment by anurse** [comment]
I'm going to close this bug just so we don't have to track it on our list of work to do :). I'm happy to continue discussing any questions you have though! Feel free to comment further to ask more questions.

**Comment by anurse** [comment]
> why is it OK to inject IHubContext into a singleton if the hub http context changes on each connection

The `IHubContext` is not connected to the `HttpContext`. It is a singleton that gives you access to all the connections currently established on the Hub. You'll notice, for example, you don't have access to the `.Caller` property on `IHubContext`, because you're not in the context of an "Invocation" so there is no caller. That's one example of how `IHubContext` is a different way of accessing a Hub. `IHubContext` may not be the best name, think of it more like a "Handle" to the Hub that gives you access to invoke methods on all clients connected to the Hub. The name comes from previous versions of SignalR so we wanted to keep it around.

> what will happen to the injected hub context when users disconnect

It'll keep working for whichever connections **are** still established.

> If we access the hub context from a singleton how will we differentiate between requests? ... For example: we want to send a message only to clients which connected to the hub with the optional parameter aaa being equal to 1

You can't. The general pattern here is to have the `OnConnectedAsync` method on your Hub put connections into named **groups** (using `.Groups.AddAsync()`) that your background services can use with `IHubContext<T>.Clients.Group("groupName")`. So you can make decisions about which group to assign a connection in the Hub, where you have access to the `HttpContext`, and then use the group to send messages.

> And another question: can a client connect to several hubs simultaneously over one connection to save connections? this would improve the performance a lot

No, this is not possible in the current version. If you're using Hubs just to separate connections into different groups, use "Groups" instead :). Otherwise, you can consider merging hubs but having them call off to different services in order to keep your application well-factored, but still use few Hubs.

> I have been going over online examples and it is unclear what is the lifespan of a signalr Hub.

A Hub is created for **each** method call. So when a client connects, we create a new instance, call `OnConnectedAsync`, then dispose the instance. Then when that client send an invocation, we create a new instance, call the method, then dispose the instance. Finally, when a client disconnects, we create a new instance, call `OnDisconnetedAsync`, then dispose the instance. This avoids having a lot of extra state hanging around accidentally. If you need to track state between invocations, you can use the `Context.Items` property bag, which **is** preserved for the whole invocation. Just be aware that obviously that will mean keeping that state in memory for the life of the connection.

> in the following example I see that he saves it in a non-static data member so it seems I am misunderstanding the lifespan of a hub

He's injecting a DI service, not storing state. The Hub is being constructed separately for each invocation but `UserInfoInMemory` is a DI singleton, so it always gets the same instance of that type. I'd recommend a pattern like that over using static fields, since it's more testable and keeps all your state contained within the DI container.

> Are there anywhere production ready code examples of correctly using signalr in various scenarios?

SignalR for ASP.NET Core is not production ready yet, it won't be ready until a release candidate or RTM release. So no, there are no examples yet, and I expect it will take some time.

**Comment by guylando** [comment]
Again, thanks for the response Andrew!

> No, this is not possible in the current version. If you're using Hubs just to separate connections into different groups, use "Groups" instead

From what I conclude from what you said, what we can\should do is to have one single hub and in the hub onConnect event to pass from the clientside a url parameter and according to it to add the client to different groups and that way we can achieve the same separation as achieved by using multiple hubs, however we will have only one single persistent http connection per user which is great. Did I get it right?


> SignalR for ASP.NET Core is not production ready yet, it won't be ready until a release candidate or RTM release. So no, there are no examples yet, and I expect it will take some time.

I would like to hear your personal opinion on this: we now use in production: Microsoft.AspNetCore.SignalR.Server on server and signalr 2.2.2 on client. And even though its old and we misused a lot of it as you now know, it still works so far pretty good in production (in a non perfect world).
Under this situation do you think it will be better to upgrade to the new signalr libraries even though they are not production ready because our current signalr version wasnt production ready too. What do you think? After talking to you I am now considering upgrading the signalr to the latest version even though its not RTM and rewriting the code as you suggested and testing this in testing environment to see if its stable and then moving this to production. Would still be interesting to hear your thoughts on the matter. Of course having an RTM version for production is the best in a perfect world but the question is what is better for our situation right now :)

Thanks!

**Comment by anurse** [comment]
> Did I get it right?

Bingo!


> Under this situation do you think it will be better to upgrade to the new signalr libraries even though they are not production ready because our current signalr version wasnt production ready too.

I would definitely advise updating. SignalR will be production-ready as part of ASP.NET Core 2.1, which is releasing in the first half of this year (so only a few months from now). We are also actively developing the latest versions, whereas the versions you are using are long out of date. There **are** fairly significant API changes you should be aware of, but I don't expect anything that will cause problems with the kind of application you've described. Take a look at this blog post which was about the **alpha** release, but is mostly still applicable to the newer **preview2** release: https://blogs.msdn.microsoft.com/webdev/2017/09/14/announcing-signalr-for-asp-net-core-2-0/

> And if to upgrade then to which version do you suggest to upgrade?

You've got it exactly. `Microsoft.AspNetCore.SignalR 1.0.0-preview2-final` and `@aspnet/signalr 1.0.0-preview2-final`

**Comment by guylando** [comment]
Thanks a lot! We will get to work now improving our code base :)

**Comment by guylando** [comment]
@anurse While upgrading to the new Microsoft.AspNetCore.SignalR I have encountered several difficulties which also couldn't find a good documentation to understand the solution regarding them.

1)I wanted to save the online users count the simplest way possible (without implementing a custom lifetime manager which was done in the chat sample) and thought to save the users in the context.Items as you suggested however I can't find how to access this storage from outside the hub.

2)Then I tried to create in-memory users storage like in the sample and wanted to store users HubConnectionContext and again encountered a difficulty that it isn't clear how to get the HubConnectionContext object from the OnConnected\OnDisconnected methods of the hub.

Will appreciate to hear your thoughts on this, thanks!

**Comment by anurse** [comment]
> thought to save the users in the context.Items as you suggested however I can't find how to access this storage from outside the hub

`Context.Items` is a **per-connection** data store, so you can't use it for counting or tracking users like this. You can simply create a singleton DI service and use it to track the data you want in the `OnConnectedAsync` method

> that it isn't clear how to get the HubConnectionContext object

This type isn't available to the Hub, as it is now a lower-level connection primitive. Why do you want to store this object? It's not designed to be stored in this way.

> I am thinking to 'hack' a simple way of getting connections count by accessing the private connections member of the hubcontext using reflection. It would be much simpler than the other solutions.

That shouldn't be necessary at all. If you want a connection count, you can use code like this:

```csharp
public class ConnectionCounterService
{
    public int ConnectionCount { get; }
    public void AddConnection()
    {
        ConnectionCount += 1;
    }
    public void RemoveConnection()
    {
        ConnectionCount -= 1;
    }
}

// In your Startup.cs ConfigureServices Method:
services.AddSingleton<ConnectionCounterService>();

// In your Hub:
public class MyHub: Hub
{
    private readonly ConnectionCounterService _counter;

    public MyHub(ConnectionCounterService counter)
    {
        _counter = counter;
    }

    public override void OnConnectedAsync()
    {
        _counter.AddConnection();
    }

    public override void OnDisconnectedAsync(Exception error)
    {
        _counter.RemoveConnection();
    }
}

// You can inject ConnectionCounterService into any other place that needs to access the connection count.
```

It's much easier for me to answer questions if you can give me context on the larger problem you are trying to solve rather than the specific types you need access to. Often we have something that works exactly for the scenario you're trying to achieve, it just isn't in the place you're looking for it :).

**Comment by guylando** [comment]
Thanks! Will take your advise.
Basically the first thing I want to achieve is updating several different locations using a single hub and a single connection for each user and having a state of all connections from every location. I think I already know how to do that using your tips above. What else I will implement will depend on what the framework will let me :) Thus I want to also ask the following: which other classes can be useful and can be injected other than IHubContext? I looked inside the DI container and saw that there are also HubLifetimeManager and IUserIdProvider and HubConnectionHandler there waiting for me. Are they maybe useful for anything? (Currently I took the user id from the http context user claim). Its good to knows what are the available tools which can help come up with interesting implementations which wouldn't come up without knowing which tools are available 😈

**Comment by anurse** [comment]
The other types are for you to **extend** SignalR, they won't really be helpful for you. You're supposed to implement/inherit from those types and add them to the container if you want to use them

* You provide a new `HubLifetimeManager` if you want to implement a new scale-out provider (`Microsoft.AspNetCore.SignalR.Redis` does this for example)
* You implement `IUserIdProvider` to set the string used to identify each connection's user when using `Clients.User(...)`
* `HubConnectionHandler` is an internal framework type you shouldn't modify or mess with. It's in DI in order to support the framework.

**Comment by guylando** [comment]
Big thanks Andrew!
Almost finished converting all our code and two questions were raised:

1) Does calling .Group().SendAsync() for groups which don't have any connections has any bad performance impact? I ask to understand if the framework handles this situation or I should check if the group is empty to not call SendAsync to improve the performance. Because we have a code which goes over all groups periodically and attempts to send message to all registered users of the groups and in most cases most of the groups wont have any users registered

2)What is the lifetime of HubCallerContext? will it get disposed if I save it in the online users singleton storage for as long as the connection is alive? the reason I want to save it there is to have access to the Abort method from outside the hub to be able to abort connections from the server using abort instead of a workaround which uses client side function.

**Comment by anurse** [comment]
> Does calling .Group().SendAsync() for groups which don't have any connections has any bad performance impact?

There shouldn't be a negative impact. Note that it's not possible to get a list of connections in a group because in a multi-server environment, each server only knows about it's own connections. We can send to groups because the server publishes the message to a shared bus (Redis) but we intentionally don't track any state there so there's no way to know how many connections are in a group.

> What is the lifetime of HubCallerContext?

The `HubCallerContext` is disposed after the task returned by your Hub method completes. Why do you need to Abort connections? That should only be used when the connection can't be shut down gracefully (by sending a message to tell the client to close, for example). There are ways to get to `Abort` from outside the `Hub`, but they are low-level and I'd like to understand exactly why and where you need to access them so we can provide the right guidance.



> Is there a difference between calling .SendAsync multiple times for different connections with different data versus grouping the connections in logical groups and calling SendAsync once per group

Do you mean a difference between `Clients.Clients(...)` and `Clients.Group(...)`? Yes, `Clients.Group` is much more efficient in multi-server scenarios since it publishes the message once, to a named Redis channel for the group, rather than `Clients.Clients` which publishes the message to a channel for each individual connection. Also, it's generally better to use `Group` when you want to refer to well-known collections of connections as it allows the system to optimize better.

**Comment by anurse** [comment]
> If you are saying that calling a client function instead and make him close the connection is better then I will do that (sounded to me like an extra step which is less efficient).

Yeah, I'd recommend doing that. An extra hop during shutdown isn't really going to negatively impact your performance, and this avoids having to deal with filtering out errors you expect from any diagnostics you want to collect. In general, a graceful shutdown is much better than using things like "Abort" in regular circumstances.
